core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,286 @@
+    case E::IS_READING:
+      return "there is already a pending read operation on this stream";
+    case E::INTERNAL_ERROR:
+      return "internal error happened";
+  }
+  return "unknown error";
+}
+
+#define SHP_FROM_PROMISE(p) \
+  std::make_shared<std::decay_t<decltype(p)>>(std::forward<decltype(p)>(p))
+
+namespace libp2p::connection {
+
+  YamuxStream::YamuxStream(std::shared_ptr<YamuxedConnection> conn,
+                           YamuxedConnection::StreamId stream_id)
+      : yamux_{std::move(conn)}, stream_id_{stream_id} {}
+
+  cti::continuable<YamuxStream::WriteResult> YamuxStream::write(
+      gsl::span<const uint8_t> in) {
+    return write(in, false);
+  }
+
+  cti::continuable<YamuxStream::WriteResult> YamuxStream::writeSome(
+      gsl::span<const uint8_t> in) {
+    return write(in, true);
+  }
+
+  cti::continuable<YamuxStream::WriteResult> YamuxStream::write(
+      gsl::span<const uint8_t> in, bool some) {
+    if (!is_writable_) {
+      return cti::make_ready_continuable(Error::NOT_WRITABLE);
+    }
+    if (is_writing_) {
+      return cti::make_ready_continuable(Error::IS_WRITING);
+    }
+    is_writing_ = true;
+
+    if (send_window_size_ - in.size() >= 0) {
+      // we can write immediately - window size on the other side allows us
+      return cti::make_continuable<WriteResult>(
+          [t = shared_from_this(), in, some](auto &&promise) {
+            auto p = SHP_FROM_PROMISE(promise);
+            t->yamux_->streamWrite(
+                t->stream_id_, in, some, [t, p = std::move(p), in](auto &&res) {
+                  t->is_writing_ = false;
+                  if (res) {
+                    t->send_window_size_ -= in.size();
+                  }
+                  p->set_value(std::forward<decltype(res)>(res));
+                });
+          });
+    }
+
+    // each time a window size gets updated, that lambda will be called, and if
+    // that time the window is wide enough, finally write to the connection
+    return cti::make_continuable<WriteResult>([t = shared_from_this(), some,
+                                               in](auto &&promise) {
+      auto p = SHP_FROM_PROMISE(promise);
+      t->yamux_->streamAddWindowUpdateHandler(
+          t->stream_id_, [t, p = std::move(p), some, in] {
+            if (t->send_window_size_ - in.size() < 0) {
+              // not yet
+              return false;
+            }
+            t->yamux_->streamWrite(
+                t->stream_id_, in, some, [t, p = std::move(p), in](auto &&res) {
+                  t->is_writing_ = false;
+                  if (res) {
+                    t->send_window_size_ -= in.size();
+                  }
+                  p->set_value(std::forward<decltype(res)>(res));
+                });
+            return true;
+          });
+    });
+  }
+
+  cti::continuable<YamuxStream::ReadResult> YamuxStream::read(size_t bytes) {
+    if (bytes == 0) {
```

Comment:
These three checks appear in the code more than one time, maybe put them to a method?